raiden/encoding/format.py

```diff
@@ -37,79 +27,107 @@ def make_field(name, size_bytes, format_string, encoder=None):
+
+    return names_slices
+
+
 def namedbuffer(buffer_name, fields_spec):  # noqa (ignore ciclomatic complexity)
-    ''' Wraps a buffer instance using the field spec.
+    ''' Class factory, returns a class to wrap a buffer instance and expose the
+    data as fields.
 
     The field spec specifies how many bytes should be used for a field and what
     is the encoding / decoding function.
     '''
+    # pylint: disable=protected-access,unused-argument
 
     if not len(buffer_name):
         raise ValueError('buffer_name is empty')
 
     if not len(fields_spec):
         raise ValueError('fields_spec is empty')
 
-    if any(field.size_bytes < 0 for field in fields_spec):
+    fields = [
+        field
+        for field in fields_spec
+        if not isinstance(field, Pad)
+    ]
+
+    if any(field.size_bytes < 0 for field in fields):
         raise ValueError('negative size_bytes')
 
-    if any(len(field.name) < 0 for field in fields_spec):
+    if any(len(field.name) < 0 for field in fields):
         raise ValueError('field missing name')
 
-    if any(count > 1 for count in Counter(field.name for field in fields_spec).values()):
-        raise ValueError('repeated field name')
-
-    size = sum(field.size_bytes for field in fields_spec)
-
-    fields = list()
-    name_slice = dict()
-    name_field = dict()
-
-    start = 0
-    for field in fields_spec:
-        end = start + field.size_bytes
+    names_fields = {
+        field.name: field
+        for field in fields
+    }
 
-        # don't create a slices and attributes for paddings
-        if not field.name.startswith('pad_'):
-            name_slice[field.name] = slice(start, end)
-            name_field[field.name] = field
-            fields.append(field.name)
+    if 'data' in names_fields:
+        raise ValueError('data field shadowing underlying buffer')
 
-        start = end
+    if any(count > 1 for count in Counter(field.name for field in fields).values()):
+        raise ValueError('repeated field name')
 
     # big endian format
     fields_format = '>' + ''.join(field.format_string for field in fields_spec)
+    size = sum(field.size_bytes for field in fields_spec)
+    names_slices = compute_slices(fields_spec)
+    sorted_names = sorted(names_fields.keys())
 
     def __init__(self, data):
-        if len(data) < size:
-            raise ValueError('data buffer is too small')
+        if len(data) != size:
+            raise ValueError('data buffer has the wrong size, expected {}'.format(size))
 
-        # XXX: validate or initialize the buffer?
-        self.data = data
+        object.__setattr__(self, 'data', data)
 
-    def __getattr__(self, name):
-        if name in name_slice:
-            slice_ = name_slice[name]
-            field = name_field[name]
+    # Intentionally exposing only the attributes from the spec, since the idea
+    # is for the instance to expose the underlying buffer as attributes
+    def __getattribute__(self, name):
```

Comment:
Exactly, we need to call `__getattribute__` to have more control over which fields are being exposed, specifically:
```
class A(object):
  a = 1
  def __getattr__(self, attr):
     return 2
assert A().a == 1
```
Meaning that a field named `size` would never be exposed because the class attribute has precedence.